optuna/integration/allennlp/_pruner.py

```diff
@@ -227,4 +245,16 @@ def on_epoch(
 
         self._trial.report(float(value), epoch)
         if self._trial.should_prune():
+            trainer._metric_tracker.should_stop_early = lambda: True  # type: ignore
+            self.pruned = True
+
+    def on_end(
+        self,
+        trainer: "GradientDescentTrainer",
+        metrics: Dict[str, Any] = None,
+        epoch: int = None,
+        is_primary: bool = True,
+        **kwargs: Any,
+    ) -> None:
+        if self.pruned:
             raise optuna.TrialPruned()
```

Comment:
This PR makes a pruning callback raise `TrialPruned` in `on_end`, instead of `on_epoch` (current implementation).
Note that this change is greatly inspired by #2849. Thank you so much for sending a awesome PR @tohmae.